src/systems/filecoin_vm/interpreter/vm_interpreter.go

```diff
@@ -24,162 +25,252 @@ const (
+		return
+	}
+
+	fromActor := inTree.GetActorState(message.From())
 	if fromActor == nil {
-		// TODO: This was originally exitcode.InvalidMethod; which is correct?
-		return inTree, _applyError(exitcode.ActorNotFound)
+		// Execution error; sender does not exist at time of message execution.
+		_applyError(inTree, exitcode.ActorNotFound, nil, nil, MinerPenaltySpec_ChargeUsedGas)
+		return
 	}
 
-	// make sure fromActor has enough money to run the max invocation
-	maxGasCost := _gasToFIL(message.GasLimit(), message.GasPrice())
-	totalCost := message.Value() + actor.TokenAmount(maxGasCost)
+	// Check sender balance.
+	gasLimitCost := _gasToFIL(message.GasLimit(), message.GasPrice())
+	totalCost := message.Value() + actor.TokenAmount(gasLimitCost)
 	if fromActor.Balance() < totalCost {
-		return inTree, _applyError(exitcode.InsufficientFunds_System)
+		// Execution error; sender does not have sufficient funds to pay for the gas limit.
+		_applyError(inTree, exitcode.InsufficientFunds_System, nil, nil, MinerPenaltySpec_ChargeUsedGas)
+		return
 	}
 
 	// make sure this is the right message order for fromActor
 	// (this is protection against replay attacks, and useful sequencing)
 	if message.CallSeqNum() != fromActor.CallSeqNum()+1 {
-		return inTree, _applyError(exitcode.InvalidCallSeqNum)
+		_applyError(inTree, exitcode.InvalidCallSeqNum, nil, nil, MinerPenaltySpec_ChargeUsedGas)
+		return
 	}
 
 	// WithActorForAddress may create new account actors
-	compTree, toActor = compTree.Impl().WithActorForAddress(message.To())
+	compTreePreSend := inTree
+	compTreePreSend, toActor := compTreePreSend.Impl().WithActorForAddress(message.To())
 	if toActor == nil {
-		return inTree, _applyError(exitcode.ActorNotFound)
+		// Execution error; receiver actor does not exist (and could not be implicitly created)
+		// at time of message execution.
+		_applyError(inTree, exitcode.ActorNotFound,
+			message.From().Ref(), minerAddr.Ref(), MinerPenaltySpec_None)
+		return
+	}
+
+	// Deduct gas limit funds from sender first.
+	// (This should always succeed, due to the sender balance check above.)
+	compTreePreSend = _withTransferFundsAssert(
+		compTreePreSend, message.From(), addr.BurntFundsActorAddr, gasLimitCost)
+
+	sendRet, compTreePostSend := _applyMessageInternal(compTreePreSend, message, vmiGasRemaining, minerAddr)
+
+	ok = _vmiUseGas(sendRet.GasUsed())
+	if !ok {
+		panic("Interpreter error: runtime execution used more gas than provided")
+	}
+
+	ok = _vmiUseGas(gascost.OnChainReturnValue(sendRet.ReturnValue()))
+	if !ok {
+		// In compTreePreSend payment for the gas limit has been transferred to burnt funds actor.
+		// When the return value serialization overflows the gas limit, pay the full balance to the miner,
+		// refunding none to the sender (even though the offending return-value won't go on-chain), and then fail.
+		_applyError(compTreePreSend, exitcode.OutOfGas,
+			addr.BurntFundsActorAddr.Ref(), minerAddr.Ref(), MinerPenaltySpec_None)
```

Comment:
This still isn't quite right. `compTreePreSend` has the sender paying gas, but the CallSeqNum is updated only in `compTreePostSend`. These two things should come together: the actor pays gas if and only if its CallSeqNum is updated.

I think this is made difficult by the CallSeqNum bump happening inside `_applyMessageInternal`: it would be easier to get right if it was pulled up here to line 192 so that both state changes happened together, or not at all.

Alternatively, the contract for _applyMessageInternal should be to return the input tree if the exit code is nonzero, and then there'll be no need to hold on to the pre-send one here, or the exit code check below.